### PR TITLE
[GPU] Fix one hot seg fault when the second input node is not a Constant node

### PR DESCRIPTION
**Pull request by yuanxion**
### Details:
 - *Get the 'depth' value from 'Select' node instead of a Constant node*

This happens when the second input of **OneHot** node is not a **Constant** node (**Select** node in this PR).
image

### Tickets:
 - *[CVS-155564](https://jira.devtools.intel.com/browse/CVS-155564)*

### CODE REVIEW

**Comment by yuanxion** [comment]
After applying the latest commit [ed76aee ](https://github.com/openvinotoolkit/openvino/pull/27442/commits/ed76aee39da9add5656505e84be78d65e40b8d03) in this PR, I can get the test_stable-fast-3d.ipynb passed now.

Here are the results comparing between CPU and GPU (without any further finetune parameters for the generated 3D result).

chair-all

**Review by jane-intel — APPROVED** [review]

File: src/core/shape_inference/include/one_hot_shape_inference.hpp [hunk 1]
```diff
@@ -67,7 +67,7 @@ std::vector<TRShape> shape_infer(const OneHot* op,
         auto depth_as_shape =
             get_input_const_data_as_shape<TRShape>(op, 1, ta, util::GetNotNegative<typename DimType::value_type>(op));
 
-        if (depth_as_shape && depth_as_shape->size() == 1) {
+        if (depth_as_shape && depth_as_shape->size() == 1 && (*depth_as_shape)[0].get_length() > 0) {
```
Comment (jane-intel) [inline src/core/shape_inference/include/one_hot_shape_inference.hpp:70]: I'm confused as to why this change in shape inference is necessary. Per the specification, depth is a positive scalar. If it is == 0, then this is an error per specification. Can you accept depth as an input in GPU?

File: src/plugins/intel_gpu/include/intel_gpu/primitives/one_hot.hpp [hunk 2]
```diff
@@ -54,6 +54,21 @@ struct one_hot : public primitive_base<one_hot> {
...
+    /// @brief onehot with depth from Select node
+    one_hot(const primitive_id& id,
+            const input_info& input,
+            const input_info& input_depth,
+            const tensor& shape,
+            const data_types output_dt,
+            const int64_t& one_hot_axis,
+            const float& on_value = 1.0f,
+            const float& off_value = 0.0f)
+        : primitive_base(id, {input, input_depth})
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/include/intel_gpu/primitives/one_hot.hpp:66]:
```suggestion
        : primitive_base(id, {input, input_depth}, 1, {optional_data_type{output_dt}})
```

File: src/core/shape_inference/include/one_hot_shape_inference.hpp (hunk 1 above)
Comment (yuanxion) [reply src/core/shape_inference/include/one_hot_shape_inference.hpp:70]: @jane-intel depth is default to 0 in one_hot primitive:
```
int64_t depth = 0; 
```
Previously it hopes to get the depth value from the second input (Constant node) of OneHot node, and then use that depth value to create one_hot primitive. But in model (in test_stable-fast-3d.ipynb), the second input of OneHot node is Select node instead of Constant node, so no depth value can be obtained during compile_model. 

```
If it is == 0, then this is an error per specification.
```
For this abnormal model, I leave the depth to be the default 0 and not a positive scalar. 

The shape_infer of one_hot will get output shape [?,0]. It will then be failed at shape_infer of the following ScatterNDUpdate node.
image
Because the updates_shape[indices_rank-1:] is 0 (updates_shape = [?, 0], which is got from shape_infer with depth = 0), and input_shape[indices_shape[-1]:] is 2 in this case, and they are not compatible. 
```
                NODE_VALIDATION_CHECK(
                    op,
                    std::all_of(inputs_shape.begin() + last_idx_dim_size, inputs_shape.end(), is_update_compatible),
                    "updates_shape[indices_rank-1:] shape must be input_shape[indices_shape[-1]:]");
```
So I changed the mentioned line code to set the updates_shape to "[?,?]" instead of "[?, 0]", and it can be compatible to 2 (input_shape[indices_shape[-1]:]) now.

Comment (jane-intel) [reply src/core/shape_inference/include/one_hot_shape_inference.hpp:70]: I don't see this model as abnormal. Depth is not guaranteed to be constant.
The fact that depth defaults to 0 in the plugin -- is an implementation detail of the plugin and plugin only. I don't see why this should affect the way common shape inference works for this operation. Especially because 0 is not a permissible value by spec.

Comment (sshlyapn) [reply src/core/shape_inference/include/one_hot_shape_inference.hpp:70]: @yuanxion, it seems we just should pass tensor_accessor with empty tensors map, which will result in a dynamic dimension for the output shape in cases like this when the actual depth value is not known yet at the moment of the shape inference call. Could you please try this approach while keeping the shape_infer function itself unchanged?

Comment (yuanxion) [reply src/core/shape_inference/include/one_hot_shape_inference.hpp]: Thanks @jane-intel  & @sshlyapn. After using empty const_data map and keeping shape_infer as it is (commit 
82b1706), it can get dynamic output_shape during compile-time. And the final 3D output result is almost the same as CPU as well.
image

File: src/plugins/intel_gpu/include/intel_gpu/primitives/one_hot.hpp (hunk 2 above)
Comment (yuanxion) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/one_hot.hpp]: Thanks, addressed with commit 82b1706.

**Review by sshlyapn — APPROVED** [review]

**Comment by sshlyapn** [comment]
build_jenkins

**Review by e-ddykim — APPROVED** [review]
